### ISSUE TITLE
Compilation failure: note: /usr/bin/ld: cannot find -lavutil

### ISSUE THREAD

**Issue by emk** [meh/rust-ffmpeg-sys#10]
Thank you for releasing an alpha version of the ffmpeg crate!

I tried building an application using ffmpeg as follows:

``` sh
git clone https://github.com/emk/substudy.git
cd substudy
git checkout ffmpeg
cargo test
```

This gave me the following error:

```
   Compiling substudy v0.1.1 (file:///home/emk/w/src/substudy)
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-m64" "-L" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/emk/w/src/substudy/target/debug/substudy-25f431e2769ace13.0.o" "-o" "/home/emk/w/src/substudy/target/debug/substudy-25f431e2769ace13" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "/home/emk/w/src/substudy/target/debug/deps/liblog-35171053ffe7ab96.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding-a635ee1112a4a115.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding_index_japanese-02d1e3db3f81af39.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding_index_simpchinese-36d98fe1ee6656fc.rlib" "/home/emk/w/src/substudy/target/debug/deps/liblibc-29adb837ec836726.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserialize-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libterm-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblog-1bf6e69c.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding_index_singlebyte-989206689effef1e.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding_index_tradchinese-1ea03524459844c3.rlib" "/home/emk/w/src/substudy/target/debug/deps/libregex-c5abedf84fe61ddb.rlib" "/home/emk/w/src/substudy/target/debug/deps/libaho_corasick-713d9a23dcd56f15.rlib" "/home/emk/w/src/substudy/target/debug/deps/libmemchr-49395bf503c0fa9c.rlib" "/home/emk/w/src/substudy/target/debug/deps/libffmpeg-9b3f15ec48a81b15.rlib" "/home/emk/w/src/substudy/target/debug/deps/libffmpeg_sys-72b52694e5f2edc4.rlib" "/home/emk/w/src/substudy/target/debug/deps/libbitflags-eec34728826d9769.rlib" "/home/emk/w/src/substudy/target/debug/deps/libregex_syntax-6d9e9f3bd76497d4.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-1bf6e69c.rlib" "/home/emk/w/src/substudy/target/debug/deps/libuchardet-19094d88b657fbb7.rlib" "/home/emk/w/src/substudy/target/debug/deps/libuchardet_sys-d6c135eb96b7863f.rlib" "/home/emk/w/src/substudy/target/debug/deps/liblibc-540159808ccfa9ab.rlib" "/home/emk/w/src/substudy/target/debug/deps/libencoding_index_korean-e6ea16a131576deb.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-1bf6e69c.rlib" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-1bf6e69c.rlib" "-L" "/home/emk/w/src/substudy/target/debug" "-L" "/home/emk/w/src/substudy/target/debug/deps" "-L" "/home/emk/w/src/substudy/target/debug/build/uchardet-sys-d6c135eb96b7863f/out/lib" "-L" "/home/emk/.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/emk/w/src/substudy/.rust/lib/x86_64-unknown-linux-gnu" "-L" "/home/emk/w/src/substudy/lib/x86_64-unknown-linux-gnu" "-Wl,-Bstatic" "-Wl,-Bdynamic" "-l" "c" "-l" "m" "-l" "avutil" "-l" "avcodec" "-l" "avdevice" "-l" "avformat" "-l" "avfilter" "-l" "avresample" "-l" "swresample" "-l" "swscale" "-l" "stdc++" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "rt" "-l" "pthread" "-l" "c" "-l" "m" "-Wl,-rpath,$ORIGIN/../../../../../.multirust/toolchains/stable/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-rpath,/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-l" "compiler-rt"
note: /usr/bin/ld: cannot find -lavutil
/usr/bin/ld: cannot find -lavcodec
/usr/bin/ld: cannot find -lavdevice
/usr/bin/ld: cannot find -lavformat
/usr/bin/ld: cannot find -lavfilter
/usr/bin/ld: cannot find -lavresample
/usr/bin/ld: cannot find -lswresample
/usr/bin/ld: cannot find -lswscale
collect2: error: ld returned 1 exit status
```

I'm running Ubuntu 14.04. My ffmpeg build directory is empty:

```
$ tree target/debug/build/ffmpeg-sys-72b52694e5f2edc4/
target/debug/build/ffmpeg-sys-72b52694e5f2edc4/
├── build-script-build
├── out
└── output

1 directory, 2 files
```

Also:

```
$ dpkg -l '*avutil*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
ii  libavutil52:am 6:9.18-0ubun amd64        Libav utility library
```

I'm not sure what's going on here, but it's probably some fairly obvious build-system issue that I'm too tired to see right now.

Thank you for wrapping this library; I can't wait to take it for a spin.

**Comment by meh** [comment]
You'll need the dev packages too, they're the ones that provide the version-less symlink.

**Comment by emk** [comment]
I'd already tried that variant, with:

``` sh
sudo apt-get install libavutil-dev
```

This gives the same error, even after `cargo clean` and a rebuild. The libavtil-dev package contains:

```
$ dpkg -L libavutil-dev
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/libavutil-dev
/usr/share/doc/libavutil-dev/copyright
/usr/include
/usr/include/libavutil
/usr/include/libavutil/avconfig.h
/usr/include/libavutil/cpu.h
/usr/include/libavutil/samplefmt.h
/usr/include/libavutil/pixdesc.h
/usr/include/libavutil/audioconvert.h
/usr/include/libavutil/imgutils.h
/usr/include/libavutil/common.h
/usr/include/libavutil/attributes.h
/usr/include/libavutil/channel_layout.h
/usr/include/libavutil/error.h
/usr/include/libavutil/rational.h
/usr/include/libavutil/audio_fifo.h
/usr/include/libavutil/lfg.h
/usr/include/libavutil/avstring.h
/usr/include/libavutil/bswap.h
/usr/include/libavutil/log.h
/usr/include/libavutil/avassert.h
/usr/include/libavutil/random_seed.h
/usr/include/libavutil/base64.h
/usr/include/libavutil/sha.h
/usr/include/libavutil/intreadwrite.h
/usr/include/libavutil/blowfish.h
/usr/include/libavutil/lzo.h
/usr/include/libavutil/mem.h
/usr/include/libavutil/fifo.h
/usr/include/libavutil/xtea.h
/usr/include/libavutil/version.h
/usr/include/libavutil/aes.h
/usr/include/libavutil/eval.h
/usr/include/libavutil/old_pix_fmts.h
/usr/include/libavutil/intfloat_readwrite.h
/usr/include/libavutil/time.h
/usr/include/libavutil/pixfmt.h
/usr/include/libavutil/file.h
/usr/include/libavutil/intfloat.h
/usr/include/libavutil/md5.h
/usr/include/libavutil/dict.h
/usr/include/libavutil/adler32.h
/usr/include/libavutil/opt.h
/usr/include/libavutil/parseutils.h
/usr/include/libavutil/mathematics.h
/usr/include/libavutil/crc.h
/usr/include/libavutil/avutil.h
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libavutil.a
/usr/lib/x86_64-linux-gnu/pkgconfig
/usr/lib/x86_64-linux-gnu/pkgconfig/libavutil.pc
/usr/share/doc/libavutil-dev/changelog.Debian.gz
/usr/lib/x86_64-linux-gnu/libavutil.so
```

There's only a few libavutil packages listed:

```
> apt-cache search avutil
libavutil-dev - Development files for libavutil
libavutil-extra-52 - Libav utility library (transitional package)
libavutil52 - Libav utility library
```

Investigation reveals that `libavutil-extra-52` is useless.

I'm not honestly sure what's up with the Ubuntu 14.04 packaging here.

**Comment by meh** [comment]
Are you 100% sure it still isn't finding avutil? Because `/usr/lib/x86_64-linux-gnu/libavutil.so` is present now.

You'll need the dev packages for every other libav thing too.

**Comment by emk** [comment]
> You'll need the dev packages for every other libav thing too.

Thank you so much, and yes, I feel like an idiot. It turns out that each of these libraries is packaged separately, and I only had some of them installed. I've manged to find _almost_ all of them. But unfortunately, one library seems to be missing under Ubuntu 14.04 LTS:

```
note: /usr/bin/ld: cannot find -lswresample
```

There is a discussion of the issue [here](http://stackoverflow.com/questions/12651816/libswresample-in-recent-ubuntu-version) which might help.

**Comment by meh** [comment]
That's easily solvable.

``` toml
[dependencies.ffmpeg]
default-features = false
features = ["codec", "format"]
```

That will only require `libavutil`, `libavformat` and `libavcodec`.

**Comment by emk** [comment]
Thank you! I had tried that, but I was missing `default-features`, and was still trying to pull in the missing library. But with `default-features = false` it works beautifully.

Thank you very much for helping me sort out my stupid mistakes. I have finally manged to link `ffmpeg` and now I'm going to try out some of the cool features you have so kindly provided. :-) Once again, I really appreciate your help.
